### PR TITLE
Register

### PR DESCRIPTION
**Pull request by NDBman**
Could test email's validity in service layer because Pattern and Matcher classes aren't mockable.

### CODE REVIEW

File: groupnews-dao/src/main/java/com/epam/internship/entity/UserEntity.java [hunk 1]
```diff
@@ -1,14 +1,30 @@
...
+@Entity
+@NoArgsConstructor
+@AllArgsConstructor
+@Table(name="users")
 public class UserEntity {
 
+	@Id
+	@GeneratedValue(strategy=GenerationType.IDENTITY)
+	private Long id;
+	@Column(nullable=false)
```
Comment (uzonyib) [inline groupnews-dao/src/main/java/com/epam/internship/entity/UserEntity.java]: nullable = false

File: groupnews-dao/src/main/java/com/epam/internship/entity/UserEntity.java [hunk 2]
```diff
@@ -1,14 +1,30 @@
...
 	private String name;
+	@Column(unique= true)
```
Comment (uzonyib) [inline groupnews-dao/src/main/java/com/epam/internship/entity/UserEntity.java]: unique = true

File: groupnews-service/src/main/java/com/epam/internship/UserService.java [hunk 3]
```diff
@@ -8,4 +8,6 @@
 
 	public User createUser(String name, String email);
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/UserService.java]: It is recommended to omit implicit public modifiers inside interfaces.

File: groupnews-service/src/main/java/com/epam/internship/UserService.java [hunk 4]
```diff
@@ -8,4 +8,6 @@
...
 	public List<User> getAllUsers();
+	public boolean emailAlreadyExists(String email);
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/UserService.java]: As discussed, do not make validation methods public.

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -1,32 +1,65 @@
...
 @Service
 public class UserServiceImpl implements UserService {
+
+	@Autowired
+	private UserRepository userRepository;
+
+	@Autowired
+	private ConversionService conversionService;
+
+	private final String EMAIL_PATTERN = "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@"
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: Please make it static.

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -1,32 +1,65 @@
...
-	private List<User> users = new ArrayList<>();
+	private Pattern pattern = Pattern.compile(EMAIL_PATTERN);
 	
-	public UserServiceImpl() {
-		users.add(User.builder().name("Mr. Brown").email("brown@email.com").build());
-	}
-
 	@Override
 	public User createUser(String name, String email) {
 		User newUser = User.builder().name(name).email(email).build();
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: Validate email here please.

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -1,32 +1,65 @@
...
+		return (List<User>) conversionService.convert(userRepository.findAll(),
+				TypeDescriptor.collection(List.class, TypeDescriptor.valueOf(UserEntity.class)),
+				TypeDescriptor.collection(List.class, TypeDescriptor.valueOf(User.class)));
+
+	}
+
+	@Override
+	public boolean emailAlreadyExists(String email) {
+		for(UserEntity userEntity : userRepository.findAll()){
+			if(userEntity.getEmail().equals(email))
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: It is recommended to use curly braces ({}) even if there is only a single statement in the body of the if / for. This improves readability a lot.

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -1,32 +1,65 @@
...
+				return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean isEmailValid(String email) {
+		Matcher matcher = pattern.matcher(email);
+		if(matcher.find()){
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: if (matcher.find()) {
Formatting & readability!

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -1,32 +1,65 @@
...
+	private UserRepository userRepository;
+
+	@Autowired
+	private ConversionService conversionService;
+
+	private final String EMAIL_PATTERN = "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@"
+			+ "[A-Za-z0-9-]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$";
 	
-	private List<User> users = new ArrayList<>();
+	private Pattern pattern = Pattern.compile(EMAIL_PATTERN);
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: Pattern can also be static.

File: groupnews-web/src/main/java/com/epam/internship/controller/UserController.java [hunk 10]
```diff
@@ -3,24 +3,40 @@
...
 	private UserService userService;
 
-	@RequestMapping(produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
-	@ResponseBody
+	@GetMapping("users")
 	public List<User> getUsers() {
 		return userService.getAllUsers();
 	}
+	
+	@PostMapping(value="new")
```
Comment (uzonyib) [inline groupnews-web/src/main/java/com/epam/internship/controller/UserController.java]: I recommend mapping this functionality to POST /users. It is more REST-like.

File: groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+@RunWith(SpringRunner.class)
+@WebMvcTest(controllers = {UserController.class})
+public class UserControllerTest {
+
+	@Autowired
+	private MockMvc mockMvc;
+	
+	@MockBean
+	private UserService systemUnderTest;
```
Comment (uzonyib) [inline groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java]: UserService is not the subject of your test, so please do not call it 'systemUnderTest'. It is misleading.

File: groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		.andDo(MockMvcResultHandlers.print())
+		//Then
+		.andExpect(MockMvcResultMatchers.status().isBadRequest());
+	}
+	
+	@Test
+	public void shouldReturnConflictStatus() throws Exception{
+		Mockito.when(systemUnderTest.emailAlreadyExists(CREATE_EMAIL)).thenReturn(true);
+		mockMvc.perform(MockMvcRequestBuilders.post("/new").param("name", CREATE_NAME).param("email", CREATE_EMAIL))
+		.andDo(MockMvcResultHandlers.print())
```
Comment (uzonyib) [inline groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java]: Is this needed?

File: groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.jsonPath("$.name").value(CREATE_NAME))
+				.andExpect(MockMvcResultMatchers.jsonPath("$.email").value(CREATE_EMAIL));
+	}
+	
+	@Test
+	public void shouldReturnBadRequestStatus() throws Exception{
+		//When
+		mockMvc.perform(MockMvcRequestBuilders.post("/new").param("name", "").param("email", CREATE_EMAIL))
+		.andDo(MockMvcResultHandlers.print())
```
Comment (uzonyib) [inline groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java]: Formatting! (Compare indentation to lines 69-73.)

**Comment by uzonyib** [comment]
Recommended: http://www.oracle.com/technetwork/java/codeconvtoc-136057.html

**Review by uzonyib — CHANGES_REQUESTED** [review]

File: groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java (hunk 8 above)
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/impl/UserServiceImpl.java]: How about simply:
return matcher.find();

File: groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java (hunk 11 above)
Comment (NDBman) [reply groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java]: Yes, you are right, on the get_user branch I already fixed it.

File: groupnews-service/src/main/java/com/epam/internship/UserService.java (hunk 4 above)
Comment (NDBman) [inline groupnews-service/src/main/java/com/epam/internship/UserService.java]: I get this error : Cannot reduce the visibility of the inherited method from UserService.
Also, should I make it private then or other modifier would be needed?

Comment (uzonyib) [reply groupnews-service/src/main/java/com/epam/internship/UserService.java]: Remove it from the interface and make it private in the implementation class.

File: groupnews-service/src/main/java/com/epam/internship/UserService.java [hunk 14]
```diff
@@ -6,8 +6,10 @@
...
 
-	public User createUser(String name, String email);
-	public List<User> getAllUsers();
-	public boolean emailAlreadyExists(String email);
-	public boolean isEmailValid(String email);
+	User createUser(String name, String email);
+
+	List<User> getAllUsers();
+
+	boolean emailAlreadyExists(String email);
```
Comment (uzonyib) [inline groupnews-service/src/main/java/com/epam/internship/UserService.java]: This should also be removed from the interface.

Comment (uzonyib) [reply groupnews-service/src/main/java/com/epam/internship/UserService.java]: Please remove this from the interface.

**Review by uzonyib — CHANGES_REQUESTED** [review]

File: groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java [hunk 15]
```diff
@@ -14,35 +14,62 @@
...
+		user3 = User.builder().name(CREATE_NAME).email(CREATE_EMAIL).build();
 	}
 
 	@Test
 	public void shouldReturnUsers() throws Exception {
+		// When
 		Mockito.when(userService.getAllUsers()).thenReturn(Arrays.asList(user1, user2));
-		System.out.println(user1.toString());
 		mockMvc.perform(MockMvcRequestBuilders.get("/users")).andDo(MockMvcResultHandlers.print())
+				// Then
```
Comment (Attila-Neuwirth) [inline groupnews-web/src/test/java/com/epam/internship/UserControllerTest.java]: Please fix indentation.

**Review by Attila-Neuwirth — CHANGES_REQUESTED** [review]
My only concern is about indentation in general.
Please set it to 4 spaces.

**Review by uzonyib — APPROVED** [review]

**Review by Attila-Neuwirth — APPROVED** [review]

**Pull request merged**